safenode/src/protocol/wallet/mod.rs

```diff
@@ -75,42 +77,51 @@ pub trait SendClient {
     ) -> Result<TransferDetails>;
 }
 
-/// A send wallet is a wallet that, in addition to the capabilities
-/// of a deposit wallet, can also send tokens to other addresses.
-#[async_trait]
-pub trait SendWallet<C: SendClient> {
-    // /// Loads a wallet from the given path.
-    // fn load_from(path: &Path) -> Self;
-    /// Creates a new wallet with the given key.
-    fn new(key: MainKey, client: C) -> Self;
+/// A wallet has an address and a balance.
+pub trait Wallet {
     /// The address of the wallet, to which others send tokens.
     fn address(&self) -> PublicAddress;
     /// The current balance of the wallet.
     fn balance(&self) -> Token;
-    /// Used to generate a new dbc id for receiving tokens.
-    fn new_dbc_address(&self) -> DbcIdSource;
-    /// Will only deposit those that are actually accessible by this wallet.
-    fn deposit(&mut self, dbcs: Vec<Dbc>);
+}
+
+/// A send wallet is a wallet that, in addition to the capabilities
+/// of a deposit wallet, can also send tokens to other addresses.
+#[async_trait]
+pub trait SendWallet: DepositWallet {
     /// Sends the given tokens to the given addresses.
     /// Returns the new dbcs that were created.
     /// Depending on the implementation of the send client, this may
     /// also register the transaction with the network.
-    async fn send(&mut self, to: Vec<(Token, PublicAddress)>) -> Result<Vec<CreatedDbc>>;
+    async fn send<C: SendClient>(
+        &mut self,
+        to: Vec<(Token, PublicAddress)>,
+        client: &C,
+    ) -> Result<Vec<CreatedDbc>>;
 }
 
 /// A deposit wallet is a wallet that can receive tokens from other wallets.
 /// It can however not send tokens to other addresses.
-pub trait DepositWallet {
-    // /// Loads a wallet from the given path.
-    // fn load_from(path: &Path) -> Self;
-    /// Creates a new wallet with the given key.
-    fn new(key: MainKey) -> Self;
-    /// The address of the wallet, to which others send tokens.
-    fn address(&self) -> PublicAddress;
-    /// The current balance of the wallet.
-    fn balance(&self) -> Token;
+pub trait DepositWallet: Wallet {
     /// Used to generate a new dbc id for receiving tokens.
     fn new_dbc_address(&self) -> DbcIdSource;
     /// Will only deposit those that are actually accessible by this wallet.
     fn deposit(&mut self, dbcs: Vec<Dbc>);
 }
+
+#[derive(serde::Serialize, serde::Deserialize)]
+pub(super) struct KeyLessWallet {
```

Comment:
Question: KeylessWallet is a DepositWallet?

Reply:
It's just a keyless wallet. It can be used for either deposits or sending.

Desposits, where you can actually read the balance, requires a key. And sending requires a key.

Reply:
What is the purpose behind KeyLessWallet?

Reply:
Right now just to have a separate structure that can be serialized to disk without the key.

Later we will be serializing ops that are appended to a log (the register), and then hydrtate the state by replaying the ops.

But in either case, the secret key is stored in a separate way.